src/lib.rs

```diff
@@ -189,8 +212,69 @@ impl<T: Translation + Clone> Mapping<T> {
         self.previous_ttbr = None;
     }
 
+    /// Checks whether the given range can be mapped or updated while the translation is live,
+    /// without violating architectural break-before-make (BBM) requirements.
+    fn check_range_bbm<F>(
+        &self,
+        range: &MemoryRegion,
+        pa: Option<PhysicalAddress>,
+        updater: &F,
+    ) -> Result<(), MapError>
+    where
+        F: Fn(&MemoryRegion, &mut Descriptor, usize) -> Result<(), ()> + ?Sized,
+    {
+        self.walk_range(
+            range,
+            &mut |mr: &MemoryRegion, d: &Descriptor, level: usize| {
+                if d.is_valid() {
+                    if !mr.is_block(level) {
+                        // Cannot split a live block mapping
+                        return Err(());
+                    }
+
+                    // Get the new flags for this descriptor by applying
+                    // the updater function to a copy and grabbing the flags
+                    let flags = {
+                        let mut dd = *d;
+                        updater(mr, &mut dd, level)?;
+                        dd.flags().ok_or(())?
+                    };
+
+                    if !flags.contains(Attributes::VALID) {
+                        // Removing the valid bit is always ok
+                        return Ok(());
+                    }
+
+                    if let Some(pa) = pa {
```

Comment:
is it possible for updater to modify the output address?

Reply:
Ah right, `Descriptor::set` is private to the crate so I guess not.

I think it might still make the code here easier to follow if you handled the `map_range` case by having the updater set the address, but up to you.

Reply:
Not sure i follow you here. The `map_range` updater does set the output address.

Reply:
ok i think i know what you mean

Reply:
Thanks, I think that's better.